src/common/snippets/src/lowered/pass/extract_loop_invariants.cpp

```diff
@@ -0,0 +1,244 @@
+#include "snippets/snippets_isa.hpp"
+#include "snippets/utils.hpp"
+
+namespace ov {
+namespace snippets {
+namespace lowered {
+namespace pass {
+
+bool ExtractLoopInvariants::run(LinearIR& linear_ir) {
+    OV_ITT_SCOPED_TASK(ov::pass::itt::domains::SnippetsTransform, "Snippets::ExtractLoopInvariants")
+    bool modified = false;
+    const auto& loop_manager = linear_ir.get_loop_manager();
+
+    auto remove_last_loop_id = [](const std::shared_ptr<Expression>& expr) {
+        auto loop_ids = expr->get_loop_ids();
+        loop_ids.pop_back();
+        expr->set_loop_ids(loop_ids);
+    };
+
+    auto get_stride_after_move_outer = [](const LoopPort& loop_port) {
+        const auto& expr_port = loop_port.expr_port;
+        const auto& layout = expr_port->get_descriptor_ptr()->get_layout();
+        const auto& shape = expr_port->get_descriptor_ptr()->get_shape();
+
+        size_t shape_dim_idx = 0;
+        if (expr_port->get_type() == ExpressionPort::Input) {
+            shape_dim_idx = utils::get_input_dim_idx(layout, loop_port.dim_idx);
+        } else if (expr_port->get_type() == ExpressionPort::Output) {
+            shape_dim_idx = utils::get_output_dim_idx(layout, loop_port.dim_idx);
+        } else {
+            OPENVINO_THROW("Unsupported expression port type!");
+        }
+
+        int64_t stride = 1;
+        for (size_t i = shape_dim_idx; i < shape.size(); ++i) {
+            if (utils::is_dynamic_value(shape[i])) {
+                return utils::get_dynamic_value<int64_t>();
+            }
+            stride *= static_cast<int64_t>(shape[i]);
+        }
+        return stride;
+    };
+
+    auto extract_from_loop = [&](const size_t& inner_loop_id) {
+        bool extracte_complete = false;  // true means all extractable exprs are extracted from this loop
+        const auto& inner_loop_info = loop_manager->get_loop_info<UnifiedLoopInfo>(inner_loop_id);
+        std::vector<std::shared_ptr<Expression>> extracted_exprs;
+        while (!extracte_complete) {
+            const auto& inner_loop_input_ports = inner_loop_info->get_input_ports();
+            if (inner_loop_input_ports.size() == 0 && inner_loop_info->get_output_ports().size() == 0) {
+                // become a empty loop, let remove it from loop_manager
+                loop_manager->remove_loop_info(inner_loop_id);
+                extracte_complete = true;
+                break;
+            }
+            bool extract_applicable = true;
+            for (size_t i = 0; i < inner_loop_input_ports.size(); i++) {  // iter loop ports
+                const auto& port_expr = inner_loop_input_ports[i].expr_port->get_expr();
+                if (std::find(extracted_exprs.cbegin(), extracted_exprs.cend(), port_expr) != extracted_exprs.cend())
+                    continue;
+                const auto& expr_input_ports = port_expr->get_input_ports();
+                extract_applicable = true;
+                if (expr_input_ports.size() == 0)
```

Comment:
What's about to create the separate function which checks for possible expr extraction?
```
bool is_extraction_applicable(const ExpressionPtr& expr, const UnifiedLoopInfoPtr& inner_loop_info) {
    const auto& expr_input_ports = expr->get_input_ports();
    if (expr_input_ports.size() == 0)
        return false;

    const auto& inner_loop_input_ports = inner_loop_info->get_input_ports();

    bool extract_applicable = true;
    for (size_t i = 0; i < expr_input_ports.size(); ++i) {  // iter expr ports
        const auto& parent = expr->get_input_port_connector(i)->get_source().get_expr();
        bool parent_scalar_with_single_consumer =
            ov::is_type<snippets::op::Scalar>(parent->get_node()) &&
            parent->get_output_port_connector(0)->get_consumers().size() == 1;

        const auto& loop_port = inner_loop_info->find_loop_port(expr_input_ports[i]);
        // expr input port is not a loop input port, then should not extract
        // expr with single scalar parent could be extracted as well, which is common.
        if (loop_port == inner_loop_input_ports.end() && !parent_scalar_with_single_consumer) {
            return false;
        }
        if (loop_port != inner_loop_input_ports.end()) {
            // after move to outside, stride in inner loop is not 1, then should not extract.
            if (get_stride_after_move_outer(*loop_port) != 1) {
                return false;
            }
        }
    }
}
```

Then you can just call in the loop:
```
            bool has_been_extraction = true;
            for (size_t i = 0; i < inner_loop_input_ports.size(); i++) {  // iter loop ports
                const auto& port_expr = inner_loop_input_ports[i].expr_port->get_expr();
                if (is_extraction_applicable(port_expr, inner_loop_info)) {
                    ....
                    has_been_extraction = true;
                    break;  // refreshed loop_input_ports, go while() to start again.
                }
            }
            // no more extractable expr in this loop.
            if (!has_been_extraction)
                extracte_complete = true;
```

Reply:
done in [c575707](https://github.com/openvinotoolkit/openvino/pull/24044/commits/c5757078d0c1dbf4e73f824cb21c35a10b6d7cfd). Thanks Alexandra.